### PR TITLE
Milvus Client

### PR DESCRIPTION
**Pull request by filip-halt**
Looking to add a wrapper around Pymilvus for easier usage.

### CODE REVIEW

**Comment by filip-halt** [comment]
fixes 1380

**Comment by xiaofan-luan** [comment]
please squash your commit record into 1 to make the commit history easier to understand.

File: pymilvus/milvus_client/defaults.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+DEFAULT_LOAD_ARGS = {
+    "partition": None,
+    "replica_number": 1,
+}
+
+DEFAULT_CONNECTION_ARGS = {
+    "alias": "default",
+}
+
+DEFAULT_SEARCH_PARAMS = {
```
Comment (xiaofan-luan) [inline pymilvus/milvus_client/defaults.py:15]: we should change all the default metric type to cosine in the future.

File: pymilvus/milvus_client/defaults.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    "IVF_FLAT": {"metric_type": "L2", "params": {"nprobe": 10}},
+    "IVF_SQ8": {"metric_type": "L2", "params": {"nprobe": 10}},
+    "IVF_PQ": {"metric_type": "L2", "params": {"nprobe": 10}},
+    "HNSW": {"metric_type": "L2", "params": {"ef": 10}},
+    "RHNSW_FLAT": {"metric_type": "L2", "params": {"ef": 10}},
+    "RHNSW_SQ": {"metric_type": "L2", "params": {"ef": 10}},
+    "RHNSW_PQ": {"metric_type": "L2", "params": {"ef": 10}},
+    "IVF_HNSW": {"metric_type": "L2", "params": {"nprobe": 10, "ef": 10}},
+    "ANNOY": {"metric_type": "L2", "params": {"search_k": 10}},
```
Comment (xiaofan-luan) [inline pymilvus/milvus_client/defaults.py:24]: remove ANNOY, and probably IVF_HNSW

File: pymilvus/milvus_client/milvus_client.py [hunk 3]
```diff
@@ -0,0 +1,801 @@
...
+
+class MilvusClient:
+    """The Milvus Client"""
+    def __init__(
+        self,
+        collection_name: str = "ClientCollection",
+        pk_field: str = None,
+        vector_field: str = None,
+        uri: str = None,
+        shard_num: int = 2,
```
Comment (xiaofan-luan) [inline pymilvus/milvus_client/milvus_client.py:30]: leave the shard_num to be empty, server side has default value of shard number

File: pymilvus/milvus_client/milvus_client.py [hunk 4]
```diff
@@ -0,0 +1,801 @@
...
+        partitions: List[str] = None,
+        consistency_level: str = "Session",
```
Comment (xiaofan-luan) [inline pymilvus/milvus_client/milvus_client.py:32]: use bounded, it's actually much faster under most cases

File: pymilvus/milvus_client/milvus_client.py [hunk 5]
```diff
@@ -0,0 +1,801 @@
...
+            self._create_collection(input_data)
+        # Nothin to init from
+        else:
+            logger.debug(
+                "No information to perform init from for collection %s",
+                self.collection_name,
+            )
+            return
+        self._create_index()
+        # Partitions only allowed on Milvus at the mmoment
```
Comment (xiaofan-luan) [inline pymilvus/milvus_client/milvus_client.py:569]: typo: moment

File: pymilvus/milvus_client/milvus_client.py [hunk 6]
```diff
@@ -0,0 +1,801 @@
...
+                If none, will use a default one. Defaults to None.
+            timeout (Optional[int], optional): What timeout to use for function calls. Defaults
+                to None.
+            drop_old (bool, optional): If a collection with the same name already exists, drop it.
+                Defaults to False.
+        """
+        self.uri = uri
+        self.collection_name = collection_name
+        self.shard_num = shard_num
+        self.partitions = partitions
```
Comment (xiaofan-luan) [inline pymilvus/milvus_client/milvus_client.py:72]: we should just ignore the partition here.

Later on user can specify the partition key here

Comment (xiaofan-luan) [reply pymilvus/milvus_client/milvus_client.py:72]: Milvus will automatially create 64 or 256 partition for the user

File: pymilvus/milvus_client/milvus_client.py [hunk 7]
```diff
@@ -0,0 +1,801 @@
...
+            self.vector_field,
+        )
+
+    def _create_index(self) -> None:
+        """Create a index on the collection"""
+        if self._get_index() is None:
+            # If no index params, use a default HNSW based one
+            if self.index_params is None:
+                # TODO: Once segment normalization we can default to IP
+                metric_type = (
```
Comment (xiaofan-luan) [inline pymilvus/milvus_client/milvus_client.py:699]: user has to be able to set default metric type I guess

Comment (xiaofan-luan) [reply pymilvus/milvus_client/milvus_client.py:699]: we are gonna to support cosine soon,so far the default set to L2 works for me but it has to been changed soon.

File: pymilvus/milvus_client/milvus_client.py [hunk 8]
```diff
@@ -0,0 +1,801 @@
...
+        pk_field: str = None,
+        vector_field: str = None,
+        uri: str = None,
+        shard_num: int = 2,
+        partitions: List[str] = None,
+        consistency_level: str = "Session",
+        replica_number: int = 1,
+        index_params: dict = None,
+        timeout: Optional[int] = None,
+        drop_old: bool = False,
```
Comment (xiaofan-luan) [inline pymilvus/milvus_client/milvus_client.py:36]: drop may not be a good idea on init, especially on multi client situation.
I would rather let the user to drop the collection on close if they'd like to

File: pymilvus/milvus_client/milvus_client.py [hunk 9]
```diff
@@ -0,0 +1,801 @@
...
+            )
+            return
+        self._create_index()
+        # Partitions only allowed on Milvus at the mmoment
+        if self.is_self_hosted and self.partitions is not None:
+            self.add_partitions(self.partitions)
+        self._create_default_search_params()
+        self._load()
+
+    def _create_collection(self, data: dict) -> None:
```
Comment (xiaofan-luan) [inline pymilvus/milvus_client/milvus_client.py:575]: We gonna to simply this part whenever the dynamic schema is ready. This is awful

File: pymilvus/milvus_client/milvus_client.py [hunk 10]
```diff
@@ -0,0 +1,801 @@
...
+        if self._get_index() is not None:
+            # Check if the collection is loaded or in progress of loading
+            if (
+                utility.load_state(
+                    self.collection_name, using=self.alias, timeout=self.timeout
+                )
+                != LoadState.NotLoad
+            ):
+                # IF the collection is loaded/loading, check the replica count
+                if len(self.collection.get_replicas().groups) == self.replica_number:
```
Comment (xiaofan-luan) [inline pymilvus/milvus_client/milvus_client.py:761]: pls open a issue in Milvus.
We have to make load idempotent.
It should dynamic add or remove one replica when ever you call load

Comment (xiaofan-luan) [reply pymilvus/milvus_client/milvus_client.py:761]: I though there is a refresh parameter which did this, pleas double check with Alex @soothing-rain

File: pymilvus/milvus_client/milvus_client.py [hunk 11]
```diff
@@ -0,0 +1,801 @@
...
+from pymilvus.orm import utility
+from pymilvus.orm.collection import Collection, CollectionSchema, FieldSchema
+from pymilvus.orm.connections import connections
+from pymilvus.orm.types import DataType, infer_dtype_bydata
+
+logger = logging.getLogger()
+logger.setLevel(logging.DEBUG)
+
+
+class MilvusClient:
```
Comment (xiaofan-luan) [inline pymilvus/milvus_client/milvus_client.py:22]: I think you will need a add a close method to:
close the connection
cleanup all the collections, index and release the data if necessary.

File: pymilvus/milvus_client/milvus_client.py (hunk 9 above)
Comment (xiaofan-luan) [reply pymilvus/milvus_client/milvus_client.py:575]: I love this code BTW, this is the best we can do right now.
Just hate the inference part because it is error prone

File: pymilvus/milvus_client/milvus_client.py [hunk 12]
```diff
@@ -0,0 +1,801 @@
...
+            for field_dict in fields.values():
+                fieldschemas.append(FieldSchema(**field_dict))
+                self.fields[field_dict["name"]] = field_dict["dtype"]
+            # Create the schema for the collection
+            schema = CollectionSchema(fieldschemas)
+            # Create the collection
+            self.collection = Collection(
+                name=self.collection_name,
+                schema=schema,
+                consistency_level=self.consistency_level,
```
Comment (xiaofan-luan) [inline pymilvus/milvus_client/milvus_client.py:661]: ignore consistency and shard thing

File: pymilvus/milvus_client/milvus_client.py [hunk 13]
```diff
@@ -0,0 +1,801 @@
...
+        pks = []
+
+
+        for k in data:
+            for key, value in k.items():
+                if key in self.fields:
+                    insert_dict.setdefault(key, []).append(value)
+
+        # Insert the data in batches
+        for i in tqdm(range(0, len(data), batch_size), disable=not progress_bar):
```
Comment (xiaofan-luan) [inline pymilvus/milvus_client/milvus_client.py:147]: I love the batch design but we need to test the corner case.
For ex. batch_size = 0

File: pymilvus/milvus_client/milvus_client.py [hunk 14]
```diff
@@ -0,0 +1,801 @@
...
+                )
+                pks.extend(res.primary_keys)
+            except MilvusException as ex:
+                logger.error(
+                    "Failed to insert batch starting at entity: %s/%s", str(i), str(len(data))
+                )
+                raise ex
+        return pks
+
+    def upsert_data(
```
Comment (xiaofan-luan) [inline pymilvus/milvus_client/milvus_client.py:171]: change to upsert implementation in the future.
comment it out the performance might be really bad due to extra delete

File: pymilvus/milvus_client/milvus_client.py [hunk 15]
```diff
@@ -0,0 +1,801 @@
...
+            logger.error("Collection does not exist: %s", self.collection_name)
+            raise ValueError(
+                "Missing collection. Make sure data inserted or intialized on existing collection."
+            )
+
+        if not isinstance(data[0], list):
+            data = [data]
+
+        return_fields = list(self.fields.keys())
+        return_fields.remove(self.vector_field)
```
Comment (xiaofan-luan) [inline pymilvus/milvus_client/milvus_client.py:258]: gonna to support return fields soon

Comment (xiaofan-luan) [reply pymilvus/milvus_client/milvus_client.py:258]: for now if user has output field requirement for vector field, I guess we need to do a get_embedding_by_PK here.

File: pymilvus/milvus_client/milvus_client.py [hunk 16]
```diff
@@ -0,0 +1,801 @@
...
+        #     self.concurrent_counter += 1
+
+        if self.collection is None:
+            logger.error("Collection does not exist: %s", self.collection_name)
+            raise ValueError(
+                "Missing collection. Make sure data inserted or intialized on existing collection."
+            )
+
+        return_fields = list(self.fields.keys())
+        return_fields.remove(self.vector_field)
```
Comment (xiaofan-luan) [inline pymilvus/milvus_client/milvus_client.py:314]: query support vector field, is there a specific reason we remove it?

File: pymilvus/milvus_client/milvus_client.py [hunk 17]
```diff
@@ -0,0 +1,801 @@
...
+
+        if self.vector_field is None:
+            logger.error(
+                "Missing vector_field, cannot infer schema for collection: %s",
+                self.collection_name,
+            )
+            raise ValueError("Missing vector_field, cannot autoinfer schema.")
+
+        try:
+            self.dim = len(data[self.vector_field])
```
Comment (longjiquan) [inline pymilvus/milvus_client/milvus_client.py:611]: The dimension calculation is correct here for float vectors, but wrong for binary vectors. In fact, we already got the field type by `infer_dtype_bydata`, which can help us to determine the dimension calculation more clear.

Comment (longjiquan) [reply pymilvus/milvus_client/milvus_client.py:611]: Btw, `self.vector_field_type` should be also added. Or should we forbid the binary vector type?

File: pymilvus/milvus_client/milvus_client.py (hunk 14 above)
Comment (longjiquan) [reply pymilvus/milvus_client/milvus_client.py:171]: agree, atomicity can't be promised if we implement the `upsert` on the client side.

File: pymilvus/milvus_client/milvus_client.py [hunk 18]
```diff
@@ -0,0 +1,801 @@
...
+        if self.collection is None:
+            logger.error("Collection does not exist: %s", self.collection_name)
+            raise ValueError(
+                "Missing collection. Make sure data inserted or intialized on existing collection."
+            )
+
+        if not isinstance(data[0], list):
+            data = [data]
+
+        return_fields = list(self.fields.keys())
```
Comment (longjiquan) [inline pymilvus/milvus_client/milvus_client.py:257]: quick question: can the `return_fields` be choosen by the users? If the users didn't specify it, then we return all fields by default.

File: pymilvus/milvus_client/milvus_client.py [hunk 19]
```diff
@@ -0,0 +1,801 @@
...
+            partition_names=partitions,
+            output_fields=return_fields,
+            timeout=timeout or self.timeout,
+        )
+
+        ret = []
+        for hits in res:
+            for hit in hits:
+                ret_dict = {x: hit.entity.get(x) for x in return_fields}
+                ret.append({"score": hit.score, "data": ret_dict})
```
Comment (longjiquan) [inline pymilvus/milvus_client/milvus_client.py:275]: We should have an identifier here which can let users know which entity belongs to which query.
For example, if the query is [q1, q2] with top4, but the server returns [entity1, entity2, entity3, entity4], which entity1 and entity2 belongs to q1 and entity3 and entity4 belongs to q2, we can't handle this case here.

File: pymilvus/milvus_client/milvus_client.py [hunk 20]
```diff
@@ -0,0 +1,801 @@
...
+            timeout=timeout or self.timeout,
+        )
+
+        # TODO: Figure out thread safety
+        # with self.concurrent_lock:
+        #     self.concurrent_counter -= 1
+
+        return res
+
+    def get_embeddings_by_pk(
```
Comment (longjiquan) [inline pymilvus/milvus_client/milvus_client.py:329]: Great wrapper for the `get_embeddings_by_pk` and `delete_by_pk`. Without this, I noticed that there are many issues in the community related to how to get vectors by pk and delete by pk. I believe this will make these developers' life easier.

File: pymilvus/milvus_client/milvus_client.py [hunk 21]
```diff
@@ -0,0 +1,801 @@
...
+                current_partitions = {
+                    partition.name for partition in self.collection.partitions
+                }
+                new_partitions = input_partitions.difference(current_partitions)
+                # If partitions need to be added, add them
+                if len(new_partitions) != 0:
+                    # Try to unload the collection, if exception already released
+                    reload = False
+                    try:
+                        self.collection.release(timeout=self.timeout)
```
Comment (longjiquan) [inline pymilvus/milvus_client/milvus_client.py:454]: Refactor this after https://github.com/milvus-io/milvus/pull/22655 has been released. Then there is no necessary to release the collection when the users tried to add a partition.

**Review by longjiquan — CHANGES_REQUESTED** [review]

File: pymilvus/milvus_client/defaults.py (hunk 1 above)
Comment (filip-halt) [reply pymilvus/milvus_client/defaults.py:15]: Yes, for now I was putting L2 as we cant enforce that the data is normalized which will create very poor results. If we end up normalizing at the segment level it may be good enough to default to.

File: pymilvus/milvus_client/milvus_client.py (hunk 6 above)
Comment (filip-halt) [reply pymilvus/milvus_client/milvus_client.py:72]: They still need to name the partitions correct? Should I instead do the partition check on each insert?

File: pymilvus/milvus_client/milvus_client.py (hunk 7 above)
Comment (filip-halt) [reply pymilvus/milvus_client/milvus_client.py:699]: The user is able to set an index at init if they want to use their own index params.

File: pymilvus/milvus_client/milvus_client.py (hunk 8 above)
Comment (filip-halt) [reply pymilvus/milvus_client/milvus_client.py:36]: On close do you mean have a init arg drop_on_close and have the __delete__ also drop the collection? Or do we want dropping to only be called when the user calls it specifically. The reason I put this in so for the use case where you are doing a lot of testing and need a fresh collection each time.

File: pymilvus/milvus_client/milvus_client.py (hunk 10 above)
Comment (filip-halt) [reply pymilvus/milvus_client/milvus_client.py:761]: We have this for refreshing. The description didnt sound like that logic but I could be wrong:
image

File: pymilvus/milvus_client/milvus_client.py (hunk 13 above)
Comment (filip-halt) [reply pymilvus/milvus_client/milvus_client.py:147]: good catch

File: pymilvus/milvus_client/milvus_client.py (hunk 14 above)
Comment (filip-halt) [reply pymilvus/milvus_client/milvus_client.py:171]: Yes, this is something that I want to replace when upsert is supported. For now most integrations and users need some kind of upsert and this is the route we are recommending, so this is just a temporary solution. Once 2.3 stable is released I will replace it. I will add a comment in docstring that emphasizes that this operation is slow and not guaranteed.

File: pymilvus/milvus_client/milvus_client.py (hunk 16 above)
Comment (filip-halt) [reply pymilvus/milvus_client/milvus_client.py:314]: Ah I thought it only supports query based on pk.

File: pymilvus/milvus_client/milvus_client.py (hunk 18 above)
Comment (xiaofan-luan) [reply pymilvus/milvus_client/milvus_client.py:263]: Agreed, by default should be only score and id, but user should be able to specify it

File: pymilvus/milvus_client/milvus_client.py (hunk 17 above)
Comment (xiaofan-luan) [reply pymilvus/milvus_client/milvus_client.py:626]: we should support for binary vector, but by default we assume it's float vector

File: pymilvus/milvus_client/milvus_client.py (hunk 10 above)
Comment (xiaofan-luan) [reply pymilvus/milvus_client/milvus_client.py:776]: I think if load is done. calling another load should just return success if collection is already loaded.

File: pymilvus/milvus_client/milvus_client.py (hunk 18 above)
Comment (filip-halt) [reply pymilvus/milvus_client/milvus_client.py:263]: I think by default it might be better to do all fields. If a user connects to an instance and doesnt know the fields, they might want to grab all of them and parse. If they want to optimize they can say which specific ones.

File: pymilvus/milvus_client/milvus_client.py (hunk 10 above)
Comment (filip-halt) [reply pymilvus/milvus_client/milvus_client.py:776]: I believe that it is idempotent, it was more for the dynamic replica count that I did this logic.

File: pymilvus/milvus_client/milvus_client.py (hunk 17 above)
Comment (filip-halt) [reply pymilvus/milvus_client/milvus_client.py]: pymilvus checks for bytestrings so thats what I did here. If its binary vector the dim is 8 * the bystring length.

**Comment by longjiquan** [comment]
@filip-halt could you please solve the code style issues, otherwise this can't be merged automatically.

File: pymilvus/milvus_client/milvus_client.py (hunk 21 above)
Comment (filip-halt) [reply pymilvus/milvus_client/milvus_client.py]: ah, woops refactored too early.
